### PR TITLE
Classification task

### PR DESCRIPTION
**Pull request by kprokofi**
What's new:

- refactoring optimizers. SAM added
- refactoring schedulers. New schedulers added
- Advanced augmentations added
- New backbones
- Lr finder in two modes added
- other changes like fixing deterministic

### CODE REVIEW

File: scripts/default_config.py [hunk 1]
```diff
@@ -2,13 +2,27 @@
...
+    cfg.lr_finder.stop_after = False
+    cfg.lr_finder.path_to_savefig = ''
 
     # model
     cfg.model = CN()
     cfg.model.name = 'resnet50'
     cfg.model.pretrained = False
     cfg.model.download_weights = True
+    cfg.model.save_chkpt = True
+    cfg.model.in_size = (224,224)
```
Comment (sovrasov) [inline scripts/default_config.py]: How this parameter is related to `cfg.data.height, cfg.data.width`?

File: scripts/default_config.py [hunk 2]
```diff
@@ -2,13 +2,27 @@
...
+    cfg.lr_finder.epochs_warmup = 2
+    cfg.lr_finder.stop_after = False
+    cfg.lr_finder.path_to_savefig = ''
 
     # model
     cfg.model = CN()
     cfg.model.name = 'resnet50'
     cfg.model.pretrained = False
     cfg.model.download_weights = True
+    cfg.model.save_chkpt = True
```
Comment (sovrasov) [inline scripts/default_config.py:24]: Why do we need this parameter?

File: tools/main.py [hunk 3]
```diff
@@ -181,6 +198,33 @@ def main():
 
     print('Building {}-engine for {}-reid'.format(cfg.loss.name, cfg.data.type))
     engine = build_engine(cfg, datamanager, models, optimizers, schedulers)
+
+    if cfg.lr_finder.enable:
+        if enable_mutual_learning:
+            print("Mutual learning is on. Lr finder will treat only the main model")
```
Comment (sovrasov) [inline tools/main.py]: .. is enabled. Learning rate will be estimated for the main model only.

File: tools/main.py [hunk 4]
```diff
@@ -181,6 +198,33 @@ def main():
 
     print('Building {}-engine for {}-reid'.format(cfg.loss.name, cfg.data.type))
     engine = build_engine(cfg, datamanager, models, optimizers, schedulers)
+
+    if cfg.lr_finder.enable:
```
Comment (sovrasov) [inline tools/main.py:220]: Print that the learning rate search is started and type of the search algorithm.

File: tools/main.py [hunk 5]
```diff
@@ -181,6 +198,33 @@ def main():
...
+        scheduler = sc_builder.build_lr_scheduler()
+
+        if enable_mutual_learning:
+            models[0], optimizers[0], schedulers[0] = model, optimizer, scheduler
+        else:
+            models, optimizers, schedulers = model, optimizer, scheduler
+        # build new engine
+        engine = build_engine(cfg, datamanager, models, optimizers, schedulers)
+
+        print("Proposed lr by LR Finder: {}".format(lr))
```
Comment (sovrasov) [inline tools/main.py]: Estimated learning rate: ... . You could use f-str here since OTE requires python >= 3.6

File: torchreid/engine/engine.py [hunk 6]
```diff
@@ -94,21 +104,22 @@ def save_model(self, epoch, save_dir, is_best=False):
...
-                                'state_dict': self.models[name].state_dict(),
-                                'epoch': epoch + 1,
-                                'optimizer': self.optims[name].state_dict(),
-                                'scheduler': self.scheds[name].state_dict(),
-                                'num_classes': self.datamanager.num_train_pids
-                            },
-                            osp.join(save_dir, name),
-                            is_best=is_best
-                        )
+            ckpt_name = osp.join(save_dir, name)
```
Comment (sovrasov) [inline torchreid/engine/engine.py]: Current version of this method is correct, but here is the outdated one. Probably this happened because of the wrong resolution of a merge conflict.

File: torchreid/utils/torchtools.py [hunk 7]
```diff
@@ -284,10 +285,9 @@ def load_pretrained_weights(model, weight_path):
     new_state_dict = OrderedDict()
     matched_layers, discarded_layers = [], []
 
-    for k, v in state_dict.items():
+    for i, (k, v) in enumerate(state_dict.items()):
```
Comment (sovrasov) [inline torchreid/utils/torchtools.py]: Why do you need `i` here?

File: torchreid/utils/tools.py [hunk 8]
```diff
@@ -136,3 +136,53 @@ def get_model_attr(model, attr):
         return getattr(model.module, attr)
     else:
         return getattr(model, attr)
+
+
+class StateCacher(object):
```
Comment (sovrasov) [inline torchreid/utils/tools.py:141]: Where is `StateCacher` used?

File: torchreid/optim/sam.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
+import torch
```
Comment (sovrasov) [inline torchreid/optim/sam.py:17]: Please add here the source and the original license statement

File: torchreid/ops/utils.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
+class EvalModeSetter:
```
Comment (sovrasov) [inline torchreid/ops/utils.py:17]: Please add the source and the original license

File: torchreid/ops/self_challenging.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
+import numpy as np
```
Comment (sovrasov) [inline torchreid/ops/self_challenging.py:17]: Please add the Intel license here

File: torchreid/models/ptcv_wrapper.py [hunk 12]
```diff
@@ -12,17 +12,45 @@
...
 from torchreid.losses import AngleSimpleLinear
 
+from functools import partial
+from pytorchcv.model_provider import _models, get_model
+
 from .common import ModelInterface
 
 __all__ = ['wrapped_models']
 
+def conv1x1(in_channels,
```
Comment (sovrasov) [inline torchreid/models/ptcv_wrapper.py]: Seems like this wrapper is also a merge mistake.

**Review by sovrasov — COMMENTED** [review]
The first bunch of comments

File: scripts/default_config.py (hunk 2 above)
Comment (kprokofi) [reply scripts/default_config.py:24]: It is convenient to run a training process for checking metrics only and do not clog up memory.

File: tools/main.py (hunk 4 above)
Comment (kprokofi) [reply tools/main.py:220]: It's done in the learning rate finder method in am_softmax.py

File: torchreid/utils/tools.py (hunk 8 above)
Comment (kprokofi) [reply torchreid/utils/tools.py:141]: For learning rate finder to keep the current state of a model and an optimizer

File: tools/convert_vgg_face_2.py [hunk 13]
```diff
@@ -1,14 +1,14 @@
...
+import cv2 as cv
+from torch.utils.data import Dataset
 
 from torchreid.data.datasets.image.lfw import FivePointsAligner
 
+import os
+import os.path as osp
+import argparse
+from tqdm import tqdm
+
```
Comment (evgeny-izutov) [inline tools/convert_vgg_face_2.py]: Why did you change the import order everywhere? According to [PEP](https://www.python.org/dev/peps/pep-0008/#imports) the default order is:
1. Standard library imports.
2. Related third party imports.
3. Local application/library specific imports.

File: torchreid/data/transforms.py [hunk 14]
```diff
@@ -815,6 +932,81 @@ def __call__(self, input_tuple):
...
+            m = np.float32(np.random.beta(self.alpha, self.alpha))
+
+            mix_img = torch.zeros_like(F.to_tensor(image))
+            mix_mask = torch.zeros_like(F.to_tensor(mask)) if mask != '' else mask
+            for i in range(self.width):
+                image_aug = image.copy()
+                mask_aug = mask.copy() if mask != '' else mask
+                depth = self.depth if self.depth > 0 else np.random.randint(1, 4)
+                # filter operations that have already been used
+                # operations = set()
```
Comment (evgeny-izutov) [inline torchreid/data/transforms.py]: dead code

File: torchreid/data/transforms.py [hunk 15]
```diff
@@ -815,6 +932,81 @@ def __call__(self, input_tuple):
...
+                for _ in range(depth):
+                    op = np.random.choice(self.train_transforms)
+                    # if op not in operations:
+                    image_aug, mask_aug = op((image_aug, mask_aug))
+                    # operations.add(op)
```
Comment (evgeny-izutov) [inline torchreid/data/transforms.py]: ditto

File: torchreid/data/transforms.py [hunk 16]
```diff
@@ -815,6 +932,81 @@ def __call__(self, input_tuple):
...
+            mix_mask = torch.zeros_like(F.to_tensor(mask)) if mask != '' else mask
+            for i in range(self.width):
+                image_aug = image.copy()
+                mask_aug = mask.copy() if mask != '' else mask
+                depth = self.depth if self.depth > 0 else np.random.randint(1, 4)
+                # filter operations that have already been used
+                # operations = set()
+                for _ in range(depth):
+                    op = np.random.choice(self.train_transforms)
+                    # if op not in operations:
```
Comment (evgeny-izutov) [inline torchreid/data/transforms.py]: ditto

File: torchreid/data/transforms.py [hunk 17]
```diff
@@ -815,6 +932,81 @@ def __call__(self, input_tuple):
...
+                # Preprocessing commutes since all coefficients are convex
+                image_aug, mask_aug = self.normalize(self.to_tensor((image_aug, mask_aug)))
+                mix_img = mix_img + ws[i] * image_aug
+                mix_mask = mix_mask + ws[i] * mask_aug if mask != '' else mask
+            prob = np.random.rand(1)[0]
+            if prob <= self.p:
+                image, mask = self.normalize(self.to_tensor((image, mask)))
+                mixed_image = (1 - m) * image + m * mix_img
+                mixed_mask = (1 - m) * mask + m * mix_mask if mask != '' else mask
+                mixed_mask = Image.fromarray(mixed_mask) if mask != '' else mask
```
Comment (evgeny-izutov) [inline torchreid/data/transforms.py]: It looks like `mixed_mask` is torch tensor but it's used like a numpy array...

File: torchreid/data/transforms.py [hunk 18]
```diff
@@ -916,17 +1114,23 @@ def build_transforms(height, width, transforms=None, norm_mean=(0.485, 0.456, 0.
...
     print('+ to torch tensor of range [0, 1]')
     transform_tr += [PairToTensor()]
     print('+ normalization (mean={}, std={})'.format(norm_mean, norm_std))
     transform_tr += [PairNormalize(mean=norm_mean, std=norm_std)]
     if transforms.random_erase.enable and transforms.random_erase.norm_image:
         print('+ random erase')
         transform_tr += [RandomErasing(**transforms.random_erase)]
-    transform_tr = Compose(transform_tr)
+    # if augmix is on - all of the above augmentations will be applyed in augmix maner
+    if transforms.augmix.enable:
```
Comment (evgeny-izutov) [inline torchreid/data/transforms.py]: Is it really necessary to add augmentations after converting images to the torch tensor? I mean it's not an intuitive behavior to mix processing numpy/PIL arrays and torch tensors.

File: torchreid/engine/engine.py [hunk 19]
```diff
@@ -205,51 +239,74 @@ def run(
             self.writer = SummaryWriter(log_dir=log_dir)
 
         # Save zeroth checkpoint
-        self.save_model(-1, save_dir)
+        if self.save_chkpt and not lr_finder:
```
Comment (sovrasov) [inline torchreid/engine/engine.py]: I think we can get rid of saving this checkpoint and the corresponding parameter.

File: torchreid/engine/engine.py [hunk 20]
```diff
@@ -205,51 +239,74 @@ def run(
...
+                    lr_finder = lr_finder
                 )
-                self.save_model(self.epoch, save_dir)
+                if self.save_chkpt:
+                    self.save_model(self.epoch, save_dir)
+
+                if lr_finder:
+                    print("epoch: {}\t top1: {}\t lr: {}".format(self.epoch,
+                                                                top1,
+                                                                self.get_current_lr()))
```
Comment (evgeny-izutov) [inline torchreid/engine/engine.py]: indent

File: torchreid/engine/engine.py [hunk 21]
```diff
@@ -205,51 +239,74 @@ def run(
...
+                lr_finder = lr_finder
             )
-            self.save_model(self.epoch, save_dir)
+            if self.save_chkpt and not lr_finder:
+                self.save_model(self.epoch, save_dir)
+            if top1 and lr_finder:
+                top1 = max(top1, top1_final)
+                print("epoch: {}\t top1: {}\t lr: {}".format(self.epoch,
+                                                                top1,
+                                                                self.get_current_lr()))
```
Comment (evgeny-izutov) [inline torchreid/engine/engine.py]: indent

File: torchreid/engine/image/am_softmax.py [hunk 22]
```diff
@@ -377,4 +416,169 @@ def _apply_batch_transform(self, imgs, obj_ids):
...
+            acc_store[lr] = top1
+
+            self.models[name].module.load_state_dict(state_cacher.retrieve("model"))
+            self.optims[name].load_state_dict(state_cacher.retrieve("optimizer"))
+            self.models[name].to(model_device)
+            set_random_seed(seed)
+
+            cur_acc = acc_store[lr]
+            # break if the results got worse
+            if (best_acc - cur_acc) >= 0.017:
```
Comment (evgeny-izutov) [inline torchreid/engine/image/am_softmax.py]: magic constant

File: tools/main.py [hunk 23]
```diff
@@ -47,17 +59,18 @@ def build_auxiliary_model(config_file, num_classes, use_gpu, device_ids=None, we
...
 
     if cfg.use_gpu:
         assert device_ids is not None
 
         model = DataParallel(model, device_ids=device_ids, output_device=0).cuda(device_ids[0])
 
-    optimizer = torchreid.optim.build_optimizer(model, **optimizer_kwargs(cfg))
-    scheduler = torchreid.optim.build_lr_scheduler(optimizer, **lr_scheduler_kwargs(cfg))
-
+    op_builder = torchreid.optim.OtimizerBuilder(model, **optimizer_kwargs(cfg))
```
Comment (evgeny-izutov) [inline tools/main.py]: Is it necessary to split creation of builder and instantiation of target object?

File: torchreid/losses/fmix.py [hunk 24]
```diff
@@ -0,0 +1,181 @@
...
+
+
+def make_low_freq_image(decay, shape, ch=1):
+    """ Sample a low frequency image from fourier space
+    :param decay_power: Decay power for frequency decay prop 1/f**d
+    :param shape: Shape of desired mask, list up to 3 dims
+    :param ch: Number of channels for desired mask
+    """
+    freqs = fftfreqnd(*shape)
+    spectrum = get_spectrum(freqs, decay, ch, *shape)#.reshape((1, *shape[:-1], -1))
```
Comment (evgeny-izutov) [inline torchreid/losses/fmix.py]: dead code

File: torchreid/engine/engine.py (hunk 19 above)
Comment (kprokofi) [reply torchreid/engine/engine.py]: Do you propose deleting the saving zeroth checkpoint?

File: scripts/default_config.py [hunk 25]
```diff
@@ -82,22 +107,32 @@ def get_default_config():
...
     cfg.train.new_layers = ['classifier']  # newly added layers with default lr
     cfg.train.base_lr_mult = 0.1  # learning rate multiplier for base layers
     cfg.train.lr_scheduler = 'single_step'
+    cfg.train.base_scheduler = ''
     cfg.train.stepsize = [20]  # stepsize to decay learning rate
     cfg.train.gamma = 0.1  # learning rate decay multiplier
     cfg.train.lr_scales = []
+    cfg.train.first_cycle_steps = 5
+    cfg.train.cycle_mult = 1.
+    cfg.train.min_lr = 1e-5
```
Comment (sovrasov) [inline scripts/default_config.py:130]: Seems that there is a duplication of warmup parameters. To resolve it I propose to remove the old multi step warmup scheduler with all the related parameters. It becomes useless since now we have a common warmup scheduler with different bases.

File: tools/main.py (hunk 23 above)
Comment (sovrasov) [inline tools/main.py]: I don't think class is actually useful here. All we need is two functions and properly forwarded kwargs.
Or just call `build_optimizer()` inline:
```python
optimizer = torchreid.optim.OtimizerBuilder(model, **optimizer_kwargs(cfg)).build_optimizer()
```

File: tools/main.py [hunk 26]
```diff
@@ -150,8 +163,13 @@ def main():
...
         extra_device_ids = [None for _ in range(len(args.extra_config_files))]
 
-    optimizer = torchreid.optim.build_optimizer(model, **optimizer_kwargs(cfg))
-    scheduler = torchreid.optim.build_lr_scheduler(optimizer, **lr_scheduler_kwargs(cfg))
+    op_builder = torchreid.optim.OtimizerBuilder(model, **optimizer_kwargs(cfg))
+    optimizer = op_builder.build_optimizer()
+    if cfg.lr_finder.enable and cfg.lr_finder.mode == 'automatic':
+        scheduler = None
+    else:
+        sc_builder = torchreid.optim.SchedulerBuilder(optimizer, **lr_scheduler_kwargs(cfg))
```
Comment (sovrasov) [inline tools/main.py]: The same thing is for scheduler builder.

File: torchreid/engine/engine.py [hunk 27]
```diff
@@ -125,12 +135,34 @@ def get_current_lr(self, names=None):
...
+                else:
+                    self.scheds[name].step()
+
+    def exit_on_plataeu(self, top1, top5, mAP):
+            name = self.get_model_names()[0]
+            current_lr = self.get_current_lr()
+
+            if current_lr == self.lb_lr:
+                current_metric = np.round(top1, 4)
+                print(self.best_metric, current_metric)
```
Comment (sovrasov) [inline torchreid/engine/engine.py]: Message consisting of two bare numbers is not informative.

File: torchreid/engine/engine.py (hunk 19 above)
Comment (sovrasov) [reply torchreid/engine/engine.py]: Yep

File: torchreid/engine/engine.py [hunk 28]
```diff
@@ -125,12 +135,34 @@ def get_current_lr(self, names=None):
...
+                if (self.best_metric >= current_metric):
+                    self.iter_to_wait += 1
+                    print(f'best metric >= current metric! iter to wait: {self.iter_to_wait}')
```
Comment (sovrasov) [inline torchreid/engine/engine.py]: This message also looks strange for a user

File: torchreid/engine/engine.py [hunk 29]
```diff
@@ -205,51 +238,74 @@ def run(
...
-                self.save_model(self.epoch, save_dir)
+                if self.save_chkpt:
+                    self.save_model(self.epoch, save_dir)
+
+                if lr_finder:
+                    print("epoch: {}\t top1: {}\t lr: {}".format(self.epoch,
+                                                                top1,
+                                                                self.get_current_lr()))
+
+            if ((self.early_stoping) and
```
Comment (sovrasov) [inline torchreid/engine/engine.py]: Too much parentheses there

File: torchreid/engine/engine.py (hunk 21 above)
Comment (sovrasov) [reply torchreid/engine/engine.py]: And why not f-string

File: torchreid/engine/image/am_softmax.py [hunk 30]
```diff
@@ -166,62 +192,71 @@ def forward_backward(self, data):
...
+            total_loss.backward(retain_graph=self.enable_metric_losses)
 
-            self.optims[model_name].step()
+            for model_name in model_names:
+                optim_name = self.optims[model_name].__class__.__name__
+                for trg_id in range(self.num_targets):
+                    if self.enable_metric_losses:
+                        ml_loss_module = self.ml_losses[trg_id][model_name]
+                        ml_loss_module.end_iteration(do_backward=False)
+                if optim_name == "SAM" and step == 1:
```
Comment (sovrasov) [inline torchreid/engine/image/am_softmax.py]: I'd rather use `isinstance` here, but it's a hack anyway. Why not to check `self.enable_sam`?

File: torchreid/engine/image/am_softmax.py [hunk 31]
```diff
@@ -377,4 +416,169 @@ def _apply_batch_transform(self, imgs, obj_ids):
...
+        print('=> Start learning rate search. Mode: {}'.format(mode))
+
+        name = self.get_model_names(None)[0]
+        model_device = next(self.models[name].parameters()).device
+        self.num_batches = len(self.train_loader)
+
+        if mode == 'automatic':
+            wd = self.optims[name].param_groups[0]['weight_decay']
+            criterion = self.main_losses[0]
+            if self.optims[name].__class__.__name__ == 'SAM':
```
Comment (sovrasov) [inline torchreid/engine/image/am_softmax.py]: The same thing with detection of SAM

File: torchreid/losses/am_softmax.py [hunk 32]
```diff
@@ -205,6 +206,15 @@ def forward(self, cos_theta, target, iteration=None):
...
                 losses = F.cross_entropy(output, target, reduction='none')
 
+            if (self.aug_type and aug_index is not None and lam is not None):
+                targets1 = torch.zeros(output.size()).scatter_(1, target.unsqueeze(1).data.cpu(), 1)
+                if self.use_gpu:
+                    targets1 = targets1.cuda()
+
+                targets2 = targets1[aug_index]
+                new_targets = targets1 * lam + targets2 * (1 - lam)
+                losses = (- new_targets * F.log_softmax(output, dim=1)).sum(dim=1)
```
Comment (sovrasov) [inline torchreid/losses/am_softmax.py:216]: Seems like tat's not completely correct since `output` is generated without knowing new_targets.
Can the argmax in new_targets and targets1 be different?

File: scripts/default_config.py [hunk 33]
```diff
@@ -18,6 +31,13 @@ def get_default_config():
     cfg.model.dropout.kernel = 3
     cfg.model.dropout.temperature = 0.2
     cfg.model.dropout.dist = 'none'
```
Comment (evgeny-izutov) [inline scripts/default_config.py]: Maybe it worth to change the name to `cfg.model.dropout_backbone` for the consistency.

File: torchreid/losses/am_softmax.py (hunk 32 above)
Comment (kprokofi) [reply torchreid/losses/am_softmax.py:216]: Fixed this.
Please, check that everything is correct now

**Comment by kprokofi** [comment]
@sovrasov , @evgeny-izutov , all issues were resolved. Please, check it.

File: torchreid/losses/am_softmax.py [hunk 34]
```diff
@@ -174,45 +174,56 @@ def forward(self, cos_theta, target, aug_index=None, lam=None, iteration=None):
                 Each position contains the label index.
             iteration (int): current iteration
         """
+        if (self.aug_type and aug_index is not None and lam is not None):
+            targets1 = torch.zeros(cos_theta.size()).scatter_(1, target.unsqueeze(1).data.cpu(), 1)
```
Comment (sovrasov) [inline torchreid/losses/am_softmax.py]: Why do you need implicitly call cpu() here? If `target` was a cuda tensor, it will cause slowdown.

File: torchreid/utils/tools.py [hunk 35]
```diff
@@ -136,3 +136,53 @@ def get_model_attr(model, attr):
...
+            self.cache_dir = tempfile.gettempdir()
+        else:
+            if not os.path.isdir(self.cache_dir):
+                raise ValueError("Given `cache_dir` is not a valid directory.")
+
+        self.cached = {}
+
+    def store(self, key, state_dict):
+        if self.in_memory:
+            self.cached.update({key: copy.deepcopy(state_dict)})
```
Comment (evgeny-izutov) [inline torchreid/utils/tools.py:158]: Does it mean that cached state is stored in gpu memory too (and we have overhead to sore double weights)?

File: torchreid/losses/am_softmax.py (hunk 34 above)
Comment (evgeny-izutov) [reply torchreid/losses/am_softmax.py]: It's better to directly place tensor on device: `targets1 = torch.zeros(cos_theta.size(), device=target.device).scatter_(1, target.detach().unsqueeze(1), 1)`

File: torchreid/models/model_store.py [hunk 36]
```diff
@@ -0,0 +1,307 @@
...
+        Path to the file.
+    ignore_extra : bool, default True
+        Whether to silently ignore parameters from the file that are not present in this Module.
+    """
+    import torch
+
+    import warnings
+    from collections import OrderedDict
+
+    pretrained_dict = torch.load(file_path)
```
Comment (evgeny-izutov) [inline torchreid/models/model_store.py]: Maybe it's better to write directly `map_location="cpu"` to exclude dead parameters on GPU from previous trainings?

File: torchreid/optim/__init__.py [hunk 37]
```diff
@@ -1,4 +1,5 @@
 from __future__ import absolute_import
 
-from .optimizer import build_optimizer
 from .lr_scheduler import build_lr_scheduler
+from .optimizer import build_optimizer
+from .sam import SAM
```
Comment (evgeny-izutov) [inline torchreid/optim/__init__.py]: no new line at the and

File: torchreid/utils/tools.py (hunk 35 above)
Comment (kprokofi) [reply torchreid/utils/tools.py:158]: Seems so, I can explicitly put a model to cpu first, then use StateCacher with "in_memory" param or use a temporary file to store weights.

Comment (evgeny-izutov) [reply torchreid/utils/tools.py:158]: Did you observed the memory consumption grow during training? Also there is possible issue with pointers to the same place in GPU memory. So, yes the option to place the model on CPU directly is reasonable.

File: torchreid/losses/am_softmax.py [hunk 38]
```diff
@@ -175,7 +175,7 @@ def forward(self, cos_theta, target, aug_index=None, lam=None, iteration=None):
             iteration (int): current iteration
         """
         if (self.aug_type and aug_index is not None and lam is not None):
-            targets1 = torch.zeros(cos_theta.size()).scatter_(1, target.unsqueeze(1).data.cpu(), 1)
+            targets1 = torch.zeros(cos_theta.size(), device=target.device).scatter_(1, target.detach().unsqueeze(1), 1)
             if self.use_gpu:
                 targets1 = targets1.cuda()
```
Comment (evgeny-izutov) [inline torchreid/losses/am_softmax.py]: If `target` is on GPU then `targets1` too. So there is no need to use `.cuda()`.

File: torchreid/losses/am_softmax.py [hunk 39]
```diff
@@ -210,7 +210,7 @@ def forward(self, cos_theta, target, aug_index=None, lam=None, iteration=None):
 
             if self.label_smooth:
                 assert not self.aug_type
-                target = torch.zeros(output.size()).scatter_(1, target.unsqueeze(1).data.cpu(), 1)
+                target = torch.zeros(output.size()).scatter_(1, target.unsqueeze(1).data, 1)
```
Comment (evgeny-izutov) [inline torchreid/losses/am_softmax.py]: `target = torch.zeros(output.size(), device=target.device).scatter_(1, target.detach().unsqueeze(1), 1)`

File: torchreid/losses/am_softmax.py [hunk 40]
```diff
@@ -210,7 +210,7 @@ def forward(self, cos_theta, target, aug_index=None, lam=None, iteration=None):
...
                 if self.use_gpu:
                     target = target.cuda()
```
Comment (evgeny-izutov) [inline torchreid/losses/am_softmax.py]: ditto

File: torchreid/engine/image/am_softmax.py [hunk 41]
```diff
@@ -377,4 +412,175 @@ def _apply_batch_transform(self, imgs, obj_ids):
             permuted_idx = torch.randperm(imgs.shape[0])
             imgs = lambd * imgs + (1 - lambd) * imgs[permuted_idx]
 
+        if self.aug_type == 'fmix':
+            r = np.random.rand(1)
+            if self.alpha > 0 and r[0] <= self.aug_prob:
+                lam, fmask = sample_mask(self.alpha, self.decay_power, self.size,
+                                        self.max_soft, self.reformulate)
+                index = torch.randperm(imgs.size(0)).to(imgs.device)
```
Comment (evgeny-izutov) [inline torchreid/engine/image/am_softmax.py]: Exclude the allocation on CPU if you want to use GPU next: `index = torch.randperm(imgs.size(0), device=imgs.device)`

File: torchreid/engine/image/am_softmax.py [hunk 42]
```diff
@@ -377,4 +412,175 @@ def _apply_batch_transform(self, imgs, obj_ids):
...
+                imgs = x1 + x2
+            else:
+                self.aug_index = None
+                self.lam = None
+
+        elif self.aug_type == 'mixup':
+            r = np.random.rand(1)
+            if self.alpha > 0 and r <= self.aug_prob:
+                lam = np.random.beta(self.alpha, self.alpha)
+                index = torch.randperm(imgs.size(0)).to(imgs.device)
```
Comment (evgeny-izutov) [inline torchreid/engine/image/am_softmax.py]: ditto

File: torchreid/engine/image/am_softmax.py [hunk 43]
```diff
@@ -377,4 +412,175 @@ def _apply_batch_transform(self, imgs, obj_ids):
...
+            else:
+                self.aug_index = None
+                self.lam = None
+
+        elif self.aug_type == 'cutmix':
+            r = np.random.rand(1)
+            if self.alpha > 0 and r <= self.aug_prob:
+                # generate mixed sample
+                lam = np.random.beta(self.alpha, self.alpha)
+                rand_index = torch.randperm(imgs.size(0)).to(imgs.device)
```
Comment (evgeny-izutov) [inline torchreid/engine/image/am_softmax.py]: ditto

File: torchreid/engine/image/am_softmax.py [hunk 44]
```diff
@@ -377,4 +412,175 @@ def _apply_batch_transform(self, imgs, obj_ids):
             permuted_idx = torch.randperm(imgs.shape[0])
             imgs = lambd * imgs + (1 - lambd) * imgs[permuted_idx]
 
+        if self.aug_type == 'fmix':
```
Comment (sovrasov) [inline torchreid/engine/image/am_softmax.py:401]: Since now we have *mix augmentation implemented, I suggest to remove `self.batch_transform_cfg.enable ...` above in this function and also the related parameters in config.

File: torchreid/engine/image/am_softmax.py [hunk 45]
```diff
@@ -377,4 +412,175 @@ def _apply_batch_transform(self, imgs, obj_ids):
...
+            acc_store[lr] = top1
+
+            self.models[name].module.load_state_dict(state_cacher.retrieve("model"))
+            self.optims[name].load_state_dict(state_cacher.retrieve("optimizer"))
+            self.models[name].to(model_device)
+            set_random_seed(seed)
+
+            cur_acc = acc_store[lr]
+            # break if the results got worse with epsilon confidence
+            if (best_acc - cur_acc) >= EPS:
```
Comment (sovrasov) [inline torchreid/engine/image/am_softmax.py:559]: I don't think such comparisons of absolute values are always correct: may be we should use a relative improvement here. 
It's just a note for the future discussion, there is no need to change the exit criterion right now.

**Review by sovrasov — APPROVED** [review]

**Review by evgeny-izutov — APPROVED** [review]
